dropshot/src/server.rs

```diff
@@ -60,62 +63,47 @@ pub struct ServerConfig {
 
 /**
  * A thin wrapper around a Hyper Server object that exposes some interfaces that
- * we find useful (e.g., close()).
- * TODO-cleanup: this mechanism should probably do better with types.  In
- * particular, once you call run(), you shouldn't be able to call it again
- * (i.e., it should consume self).  But you should be able to close() it.  Once
- * you've called close(), you shouldn't be able to call it again.
+ * we find useful.
  */
 pub struct HttpServer {
     app_state: Arc<DropshotState>,
-    server_future: Option<BoxFuture<'static, Result<(), hyper::Error>>>,
+    server: Server<AddrIncoming, ServerConnectionHandler>,
     local_addr: SocketAddr,
-    close_channel: Option<tokio::sync::oneshot::Sender<()>>,
 }
 
 impl HttpServer {
     pub fn local_addr(&self) -> SocketAddr {
         self.local_addr
     }
 
-    pub fn close(mut self) {
-        /*
-         * It should be impossible to close a channel that's already been closed
-         * because close() consumes self.  It should also be impossible to fail
-         * to send the close signal because nothing else can cause the server to
-         * exit.
-         */
-        let channel =
-            self.close_channel.take().expect("already closed somehow");
-        channel.send(()).expect("failed to send close signal");
-    }
-
-    /*
-     * TODO-cleanup is it more accurate to call this start() and say it returns
-     * a Future that resolves when the server is finished?
+    /**
+     * Begins execution of the underlying Http server.
      */
-    pub fn run(&mut self) -> tokio::task::JoinHandle<Result<(), hyper::Error>> {
-        let future =
-            self.server_future.take().expect("cannot run() more than once");
-        tokio::spawn(async { future.await })
-    }
+    pub fn run(self) -> RunningHttpServer {
```

Comment:
Per the TODO that was here, and especially if we call this thing `HttpServerStarter`, I think this function is more accurately called `start()` now.